src/DirectGateway.php

```diff
@@ -43,6 +44,16 @@ public function setSimulatorMode($value)
         return $this->setParameter('simulatorMode', $value);
     }
 
+    public function getReferrerId()
+    {
+        return $this->getParameter('referrerId');
+    }
+
+    public function setReferrerId($value)
+    {
+        return $this->setParameter('referrerId', $value);
+    }
+
```

Comment:
I'll tell you the reasoning behind it - I may be totally wrong with the implementation.

I saw the ReferrerID (or referrerId) as related to the account rather than a specific transaction. So putting it at the top gateway level allowed it to be set up during the main gateway initialisation. Some plugins, such as the ignited/laravel-gateway, wrap up all the account-related information in the app config, so don't have to deal with it when processing transactions. Without the methods at that level, it cannot be set there.

If you want to set it at the action level instead, then the get/set methods also need to be there. I can see different implementations wanting to use different approaches. OmniPay kind of joins the two levels together in its internals, so you can set it at either level, and it will be available in the authorisation message.

That is all my understanding and reasons for doing it that way, any of which may be incorrect, or just not wanted. The tests, yeah, I overlooked the gateway level test.